# Homework 4: Functions and Probability (50 points)

**Reading**: Textbook chapter [9](https://inferentialthinking.com/chapters/09/Randomness.html). Section 9.4 on the Monty Hall problem is a fun read.

Please complete this notebook by filling in the cells provided. Before you begin, execute the following cell to load the needed functions. Each time you start your server, you will need to execute this cell again to load them.  

Homework 4 is due Wednesday, February 9, at 9:00am. Late work will not be accepted without prior arrangements.  

In [1]:
# Don't change this cell; just run it. 
import numpy as np
import pandas as pd
import math as m
import warnings
warnings.simplefilter('ignore', FutureWarning)

# These lines do some fancy plotting magic.\n",
import matplotlib
%matplotlib inline
import matplotlib.pyplot as plots
plots.style.use('fivethirtyeight')

## 1. Working with Text using Functions


Nellie Bly (1864-1922) was an American journalist and travel writer.  We will load an essay of hers *Ten Days in a Mad-House* and count the number and length of words in the essay.  

**Aside:  metadata**  Can you spot the metadata attached to Project Gutenberg's digiziation of Bly's investigation of asylum conditions?  https://www.gutenberg.org/ebooks/59899  There is one important, essential piece of metadata that is missing, probably because the printed book did not contain this data item.  Can you spot it?

This cell loads Bly's essay *Ten Days in a Mad-House; or, Nellie Bly's Experience on Blackwell's Island*.  The file is 156kb; it easily fits in memory at once.


In [2]:
# Just run this cell to load the text. Can you tell what the code is doing?
chapters = open("madhouse.csv").read().replace('\n',' ')
chapters[0:400] #this shows the first 400 characters in chapters starting at index 0

'                         TEN DAYS IN A MAD-HOUSE.                                    CHAPTER I.                           A DELICATE MISSION.   On the 22d of September I was asked by the _World_ if I could have myself committed to one of the asylums for the insane in New York, with a view to writing a plain and unvarnished narrative of the treatment of the patients therein and the methods of manag'

**Question 1.** The essay has 17 chapters.  Split the data into chapters, putting the text of each
chapter in a convenient datastructure.  (list or DataFrame will both work)  
Make sure the 0th item in your list is the first chapter, and make sure there are 17 chapters. 

In [3]:
# code here
# This helps split all the chapters apart from 1-17
# This makes a list
split_chapters = chapters.split("CHAPTER")[1:]

**Question 2.** Write a function called `word_count` that takes a single argument, a string (such as the text of a single chapter), and returns the number of words in that string (chapter).  Assume that words are separated from each other by spaces. 

*Hint:* Try the string method [`split`](https://docs.python.org/3/library/stdtypes.html#str.split) and the function [`len`](https://docs.python.org/3/library/functions.html#len) (clicking on these will take you to help pages). Recall that we used them in the notebook for Lecture 2.

In [4]:
# code here
def word_count(a_string):
    """This counts the number of words in a given string"""
    word_list = a_string.split()
    number_of_words = len(word_list)
    return number_of_words

num_words = word_count(split_chapters[0])
num_words

978

**Question 3 (10 points).** Count the number of words and the number of characters in each chapter and put the results in a dataframe.  Calculate and store the average word length.  (Remember, length includes the length of the spaces.  Find the average length of a word in the text). 

In [5]:
# Counting the number of words in each chapter
# Test word count
word_count_chap = [word_count(i) for i in split_chapters]
word_count_chap

[978,
 649,
 3491,
 2492,
 1036,
 3121,
 1883,
 1172,
 928,
 820,
 2691,
 2210,
 1459,
 2921,
 867,
 1325,
 1110]

In [6]:
# Counting the number of characters in each chapter
char_count = [len(i) for i in split_chapters]
char_count

[5274,
 3460,
 18653,
 13229,
 5279,
 16091,
 9868,
 6395,
 5021,
 4410,
 14039,
 11684,
 7852,
 15026,
 4731,
 7044,
 5923]

In [7]:
# Finding the average word length and putting it into dataframe
d = {'Number of Words': word_count_chap, 'Number of Characters': char_count}
df = pd.DataFrame(data=d)
df

Number of Words  Number of Characters
0               978                  5274
1               649                  3460
2              3491                 18653
3              2492                 13229
4              1036                  5279
5              3121                 16091
6              1883                  9868
7              1172                  6395
8               928                  5021
9               820                  4410
10             2691                 14039
11             2210                 11684
12             1459                  7852
13             2921                 15026
14              867                  4731
15             1325                  7044
16             1110                  5923

In [8]:
# Finding the average word length
mean_words = sum(char_count)/sum(word_count_chap)
mean_words_trunc = m.trunc(mean_words) #truncated because you can't have a word length that aren't integers
print("The average word length in the essay is", mean_words_trunc, "characters long")

The average word length in the essay is 5 characters long


## 2. False positives ##
In "Performance and Implementation Evaluation of the Abbott BinaxNOW Rapid Antigen Test in a High-Throughput Drive-Through Community Testing Site in Massachusetts"  https://journals.asm.org/doi/10.1128/JCM.00083-21  Pollack et al. test a rapid antigen test for COVID-19 spike protein at a clinic in Lawrence, MA.  For asymptomatic adults (n=974) they reported a sensitivity of 70.2 and a specificity of 99.6% in a population with a 5.9% infection rate.


**Question 4.** Suppose you have 7,000 students returning from traveling over the new year and suppose you have an infection rate like the university-wide surveillance rate the week of Jan 12 of 6.3%.
You give all of them the antigen test.  What fraction do you expect will get a false negative test (and will go to classes without quarrantine)?  What fraction will get a false negative test?


In [13]:
# code here
false_neg_frac = .059 * 7000 * .298 / 7000
false_neg_frac

0.017582

The fraction of students that will get a false negative is 0.017582 students

**Question 5.** Suppose these students go traveling again over spring break, but come back with only 1% of them infected this time.  Same questions: what is the expected number who are infected but think otherwise, and what is the expected number who get positive tests but aren't infected.



In [12]:
false_neg_pop = .01 * 7000 * .298
print(false_neg_pop)
false_pos_pop = 0.01 * 7000 * 0.004
print(false_pos_pop)

20.86
0.28


**Question 6.** 
Assuming that 2% of the student population has an infection and doesn't know it, estimate the probability of exactly 0, 1 ,2 ,3 ,4 ,or 5 such infected students in a class of 50.   You may use math or simulation as long as you get the right answers.

*Answer here*

In [40]:
import scipy.stats
from scipy.stats import binom

k= np.arange(1,5+1)
binoprob = scipy.stats.binom.pmf(k, 50, p=.02)
binoprob
print("There would be an expected propability of", binoprob, "for 1, 2, 3, 4, and 5 infected students in a class", \
     "of 50.")

There would be an expected propability of [0.37160171 0.18580086 0.06066967 0.01454834 0.00273152] for 1, 2, 3, 4, and 5 infected students in a class of 50.


*Answer here*

## 3. de Méré's problem ##

The Chevalier de Méré was a gambler, and found two propositions involving dice that had close to 50% probabilities:

**Event A**  In 4 rolls of the dice, six comes up at least once. 

**Event B**  In 24 rolls of the dice, six comes up at least twice.

He found out (by losing repeatedly) which of these was the winner, and he asked Blaise Pascal to look into the matter.


**Question 7.**
Write a function that simulates event A, returning True or False.

In [14]:
dice = [1, 2, 3, 4, 5, 6]

def six_prob():
    """Calculates the probability that 6 comes up in 4 dice rolls"""
    result = np.random.choice(dice, 4)
    result_list = result.tolist()
    count = result_list.count(6)
    if count >= 1:
        return True
    else:
        return False

six_prob()

False

**Question 8.**
Write a function that simulates event B, returning True or False.

In [15]:
dice = [1, 2, 3, 4, 5, 6]

def twentyfour_prob():
    """Calculates the probability that 6 comes up twice in 24 dice rolls"""
    result = np.random.choice(dice, 24)
    result_list = result.tolist()
    count = result_list.count(6)
    if count >= 2:
        return True
    else:
        return False

twentyfour_prob()

True

**Question 9.**
Estimate the frequencies of events A and B by running your functions 10,000 times.

In [29]:
## 4. Biased coin ##
a_rolls = [six_prob() for x in range(1, 10000+1)] # runs function 10000 times, starting at 1 to 10000
six_freq = a_rolls.count(True) / 10000 #takes frequency by dividing number of true by 10000
six_freq

0.5243

In [30]:
b_rolls = [twentyfour_prob() for x in range(1, 10000+1)] # runs function 10000 times
twentyfour_freq = b_rolls.count(True) / 10000 #takes frequency by dividing number of true by 10000
twentyfour_freq

0.9277

**Question 10.**
Draw a random number from the normal distibution centered at .0512 with standard deviation .003.    Assign this to `heads_probability`.

In [18]:
# Draw random samples from a normal (Gaussian) distribution.
heads_probability = np.random.normal(0.0512, 0.003)
heads_probability

0.05424044131890274

**Question 11.**
Write a function that simulates 1000 flips of a coin whose probability of heads is `heads_probability`

In [21]:
def biased_flipper_1000(heads_probability):
    """Simulates 1000 coin tosses with heads_probability for the chance of landing on heads"""
    return np.random.binomial(1000, heads_probability, 1)

# The function returns a list of samples
#from a binomial distribution based on the inputted parameters when calling np.random.binomial(n, p, size).
#n = number of experiments, p = probability of success, size = number of randomly selected experiments

# Test
biased_heads = biased_flipper_1000(heads_probability)
biased_heads

array([50])

**Question 12.**
Estimate the long-term proability of the event 
"Out of 1000 tosses of the biased coin, there were 500 or fewer heads."
by running your function 10,000 times.

In [34]:
tosses_array = [biased_flipper_1000(heads_probability) for x in range(1, 10000)]
tosses_array

[array([53]),
 array([62]),
 array([60]),
 array([40]),
 array([56]),
 array([51]),
 array([68]),
 array([51]),
 array([46]),
 array([50]),
 array([43]),
 array([54]),
 array([51]),
 array([65]),
 array([45]),
 array([50]),
 array([56]),
 array([59]),
 array([60]),
 array([47]),
 array([54]),
 array([51]),
 array([64]),
 array([54]),
 array([60]),
 array([53]),
 array([46]),
 array([65]),
 array([44]),
 array([58]),
 array([41]),
 array([49]),
 array([51]),
 array([58]),
 array([49]),
 array([46]),
 array([58]),
 array([68]),
 array([53]),
 array([46]),
 array([64]),
 array([55]),
 array([55]),
 array([58]),
 array([61]),
 array([59]),
 array([54]),
 array([54]),
 array([57]),
 array([43]),
 array([50]),
 array([52]),
 array([53]),
 array([59]),
 array([50]),
 array([53]),
 array([60]),
 array([54]),
 array([49]),
 array([50]),
 array([54]),
 array([42]),
 array([67]),
 array([60]),
 array([45]),
 array([54]),
 array([53]),
 array([57]),
 array([38]),
 array([42]),
 array([63]),
 array

In [ ]:
print("The long-term probability will always be 1")